Documents/UpgradeExtensibility.md

```diff
@@ -0,0 +1,157 @@
+VFSForGit Upgrade Extensibility Design Document
+===============================
+
+Scope
+-----
+
+Enable groups to extend the Upgrade verb to install / update extra
+components besides just VFSForGit and Git.
+
+Motivation
+----------
+
+VFSForGit can exist as a component of a larger engineering
+system. Some environments could include components that are coupled
+with specific versions of VFSForGit and should be updated with with
+VFSForGit. There could be other interactions specific to the
+environment that need to be accounted for when updating VFSForGit. As
+these requirements are specific to the individual environments
+VFSForGit is being deployed into, we need a way for groups to include
+extra components and logic in the VFSForGit upgrade.
+
+Requirements
+------------
+
+* Groups can include extra components / logic in VFSForGit Upgrade
+
+* Feed needs to support advertising current available version
+
+Non-Goals
+---------
+
+* Initial configuration of the alternate feed URL
+
+* Controlling the entire package of components
+
+Design
+------
+
+VFSForGit will support an alternative "Feed" from which updates can be
+retrieved. When an alternative feed is configured, VFSForGit will
+query this feed for updates instead of the normal feed. This
+alternative feed can contain customized components in addition to the
+VFSForGit and Git installers.
+
+### Feed
+
+The alternative feed will be a NuGet feed. The feed URL is specified
+in the gvfs configuration file (gvfs.config) with the following entry:
+
+    upgrade.feedurl
+
+This is similar to how upgrade rings are specified. The upgrade.feed
+will take precedence over the upgrade ring configuration. When this
+value is configured, VFSForGit will query and download updates from
+this feed. It must be a NuGet feed.
+
+### Upgrade Package Contents
+
+If using the alternative feed, the upgrade package must include
+installers for Git and VFSForGit. It can optionally include installers
+to run before / after installing Git and GVFS. The upgrade logic will
+look in the following locations for the pre / post setup installers:
+
+    PreSetup\install.exe :: Run before the Git and VFSForGit
+    			    installers are run.
+
+    PostSetup\install.exe :: Run after the Git and VFSForGit
+    			     installers are run.
+
+This means the top level structure of the NuGet package will contain
+the following directories:
+
+  G4W :: Contains Git for windows installer
+  
+  GVFS :: Contains the VFSForGitInstaller
+  
+  PreSetup :: (Optional) Contains the installer to run before setup
+  
+  PostSetup :: (Optional) Contains the installer to run after setup
+

```

Comment:
We're basically creating a vehicle for installing anything on a developer box.  How will we ensure this isn't abused?